javascript/computer_science/time_complexity.md

```diff
@@ -0,0 +1,326 @@
+
+Looking up something in one step is as good as it gets for time complexity.
+
+While we're looking at the simplest form of Big O, let's take a look at one of its little gotchas to keep in mind. You may have thought a moment ago, is it really just one step? The answer is technically no, in reality the computer must first look up where the array is in memory, then from the first element in the array it needs to jump to the index argument provided. That's at least a couple of steps. So you wouldn't be wrong for writing something like `O(1 + 2(steps))`. However, the 2 steps are merely incidental. With an array of 10,000 elements it still takes the same amount of steps as if the array was 2 elements. Because of this, Big O doesn't concern itself with these incidental numbers. They don't provide any context to how the complexity grows when the data size changes, because they are constant, and so in Big O they are dropped. Big O only wants to tell us an algorithm's complexity relative to the size of the input.
+
+Do the number of steps matter? Yes, they might. We'll touch on when this may be the case a little later.
+
+#### O(log N) - Logarithmic Complexity
+
+Logarithmic Complexity tells us that the numbers of steps an algorithm takes increases by 1 as the data doubles. That's still pretty efficient when you think about it. Going from 5,000 to 10,000 data elements and only taking one additional step can scale really well.
+
+One such algorithm that does this is Binary Search. It only works on sorted arrays but if you have an array of 10 items in sorted order 
+
+~~~js
+arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+~~~
+
+and wanted to know if it had the number `7`, Binary Search would guess the middle item of the array and see what is there. Because the array is sorted, if the number at the middle index was `6`, then we know anything to the left of that index cannot be the number 7, as those items must be lower than 6 in a sorted array.
+
+~~~js
+arr = [-, -, -, -, -, 6, 7, 8, 9, 10]
+~~~
+
+Therefore in just one step we've eliminated half of the array. We can do the same with the remaining half. We can guess the middle index and see if it's 7. Half of that (half of an array) array eliminated again. In this case the middle index would be 8 and we know that 7 is less than 8 so we can eliminate anything to the right of the number 8.
+
+~~~js
+arr = [6, 7, 8, -, -]
+~~~
+
+We can keep doing this until we have an array of just one item. If it matches the number we're looking for we've found it, if not then it isn't in the array.
+
+The below table summarises the size of an array doubling and how many steps in Big O terms we would need to arrive at one element to see if it matches what we're looking for
+
+| Size | Steps |
+| ----------- | ----------- |
+| 1 | 0 |
```

Comment:
I think in the Ruby part I made a mistake here that's carried over. I think size 1 would be 1 step. size 2 would be 2 then 3 would be 4. Not sure what I was thinking but I don't think an alogrithm can have zero steps unless it's empty. Sorry dude. I'll get it updated in the Ruby lesson